tests/RobotFramework/tests/tedge_agent/workflows/custom_restart.toml

```diff
@@ -0,0 +1,23 @@
+operation = "controled_restart"
+
+[init]
+script = "/etc/tedge/operations/log-restart.sh ${.topic.cmd_id} ${.payload.status}"
+next = ["scheduled"]
+
+[scheduled]
+script = "restart"
+next = ["restarting", "successful_restart", "failed_restart"]
```

Comment:
I still don't fully get the idea behind the need for that first `restarting` state in this array. Why not just model it as an independent `restarting` state transitioning from the `scheduled` state? That `restarting` state can then have subsequent `successful_restart` and `failed_restart` states?

Reply:
The shortcut I took I indeed confusing. I should not have hijack the __next__ field and I will fix that.

What you describe is indeed the correct workflow: moving from *scheduled* to *restarting* and then to *successful_restart* or *failed_restart*.

The actual purpose is different: these 3 state names are used as alias to the state names of the *restart* internal workflow. The states of the latter cannot be used unchanged (as the main workflow already has *executing*,  
*successful* and *failed* states) nor hard-coded (as the main workflow might invoke many restarts).

Reply:
Why is it wrong to model the same workflow as follows (note the updated definition of the `scheduled` state):

```toml
operation = "controlled_restart"

[init]
script = "/etc/tedge/operations/log-restart.sh ${.topic.cmd_id} ${.payload.status}"
next = ["scheduled"]

[scheduled]
inbuilt-action = "restart"
next = ["successful_restart", "failed_restart"]

[successful_restart]
script = "/etc/tedge/operations/log-restart.sh ${.topic.cmd_id} ${.payload.status}"
next = ["successful"]

[failed_restart]
script = "/etc/tedge/operations/log-restart.sh ${.topic.cmd_id} ${.payload.status} ${.payload.restartError}"
next = ["failed"]
```

When the `tedge-agent` processes the `inbuilt-action = restart`, it knows that it was called in the context of the `controlled_restart` command, from the `scheduled` state. It can persist all this information before the restart and once the restart completes, it can just retrieve the state from which `restart` was invoked, which is `scheduled` in this case, and then just transition to either of the states defined in its `next` field for that state.

This is simplistic view that I have about handling restart. I'm sure I'm missing something here. I'm just not able to see what it is yet.

Reply:
> When the `tedge-agent` processes the `inbuilt-action = restart`, it knows that it was called in the context of the `controlled_restart` command, from the `scheduled` state. It can persist all this information before the restart and once the restart completes, it can just retrieve the state from which `restart` was invoked, which is `scheduled` in this case, and then just transition to either of the states defined in its `next` field for that state.

With one exception described bellow, what you describe is what is actually implemented. The context of the restart (.i.e. the whole MQTT message describing the state of the main workflow is stored by the restart operation and resumed after the reboot).

> This is simplistic view that I have about handling restarts in the context of other operations by simply adding more states. I'm sure I'm missing something here. I'm just not able to see what it is yet.

Moving from a state triggering a restart to a state waiting for the end of the restart is required to avoid a race condition.

With your example, where there is no *restarting* state, there is a risk for an infinite loop. If the retained message for the *sheduled* state is read before the restart completion is detected (these are two independent threads), then the workflow will erroneously trigger a new restart.

Reply:
> With your example, where there is no restarting state, there is a risk for an infinite loop. If the retained message for the sheduled state is read before the restart completion is detected (these are two independent threads), then the workflow will erroneously trigger a new restart.

Yeah, now the need for that `restarting` state is clear. I've documented some of my thoughts on how to make this flow clearer in this [comment](https://github.com/thin-edge/thin-edge.io/pull/2479#discussion_r1409229947).